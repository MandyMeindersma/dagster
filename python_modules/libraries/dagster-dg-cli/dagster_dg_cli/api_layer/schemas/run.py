"""Run models for REST-like API."""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel


class RunStatus(str, Enum):
    """Run status enum for FastAPI compatibility."""

    QUEUED = "QUEUED"
    NOT_STARTED = "NOT_STARTED"
    MANAGED = "MANAGED"
    STARTING = "STARTING"
    STARTED = "STARTED"
    SUCCESS = "SUCCESS"
    FAILURE = "FAILURE"
    CANCELING = "CANCELING"
    CANCELED = "CANCELED"


class Run(BaseModel):
    """Run resource model."""

    id: str  # Run ID (UUID string)
    run_id: str  # Same as id, kept for compatibility
    status: RunStatus
    job_name: str
    pipeline_name: Optional[str] = None  # Legacy field
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    update_time: Optional[datetime] = None
    creation_time: datetime
    can_terminate: bool = False

    class Config:
        from_attributes = True  # For future ORM compatibility


class RunList(BaseModel):
    """GET /api/runs response."""

    items: list[Run]
    total: int
    has_next_page: bool = False
    cursor: Optional[str] = None
