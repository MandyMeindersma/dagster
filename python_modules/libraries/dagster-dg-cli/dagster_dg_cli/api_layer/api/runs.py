"""Run endpoints - REST-like interface."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from dagster_dg_cli.api_layer.graphql_adapter.run import get_run_via_graphql, list_runs_via_graphql
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.run import Run, RunList


@dataclass(frozen=True)
class DgApiRunApi:
    client: IGraphQLClient

    def list_runs(
        self,
        limit: Optional[int] = None,
        status: Optional[str] = None,
        job_name: Optional[str] = None,
        cursor: Optional[str] = None,
    ) -> "RunList":
        """List runs with optional filtering.

        Args:
            limit: Maximum number of runs to return
            status: Filter by run status
            job_name: Filter by job name
            cursor: Pagination cursor

        Returns:
            RunList: List of runs matching the filters
        """
        return list_runs_via_graphql(
            client=self.client,
            limit=limit,
            status=status,
            job_name=job_name,
            cursor=cursor,
        )

    def get_run(self, run_id: str) -> "Run":
        """Get a specific run by ID.

        Args:
            run_id: The run ID to fetch

        Returns:
            Run: The requested run

        Raises:
            Exception: If run not found or API error occurs
        """
        return get_run_via_graphql(client=self.client, run_id=run_id)
