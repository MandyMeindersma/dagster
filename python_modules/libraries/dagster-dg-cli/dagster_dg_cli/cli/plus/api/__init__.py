"""API commands for interacting with Dagster Plus REST-like interface."""

import click
from dagster_dg_core.utils import DgClickGroup

from dagster_dg_cli.cli.plus.api.deployments import deployments_group


@click.group(
    name="api",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "deployments": deployments_group,
    },
)
def api_group():
    """Make REST-like API calls to Dagster Plus."""
