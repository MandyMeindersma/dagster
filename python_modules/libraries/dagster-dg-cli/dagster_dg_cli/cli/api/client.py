"""Client factory for DG API commands."""

from typing import Protocol

import click
from dagster_shared.plus.config import DagsterPlusCliConfig

from dagster_dg_cli.utils.plus.gql_client import DagsterPlusGraphQLClient


class GraphQLClientFactory(Protocol):
    """Protocol for GraphQL client factories used in testing."""

    def __call__(self, config: DagsterPlusCliConfig) -> DagsterPlusGraphQLClient: ...


class DgApiTestContext:
    """Test context for DG API commands."""

    def __init__(self, client_factory: GraphQLClientFactory):
        self.client_factory = client_factory


def create_dg_api_graphql_client(
    ctx: click.Context, config: DagsterPlusCliConfig
) -> DagsterPlusGraphQLClient:
    """Create GraphQL client for DG API commands.

    This is the single entry point for GraphQL client creation in DG API commands.
    It checks for test context injection and falls back to creating a real client.

    Args:
        ctx: Click context from the CLI command
        config: Optional config to use, falls back to global config if not provided

    Returns:
        DagsterPlusGraphQLClient instance
    """
    # Check if we have a test context with custom factory
    if ctx.obj and isinstance(ctx.obj, DgApiTestContext) and ctx.obj.client_factory:
        return ctx.obj.client_factory(config)

    # Normal operation: create real client from config
    return DagsterPlusGraphQLClient.from_config(config)
