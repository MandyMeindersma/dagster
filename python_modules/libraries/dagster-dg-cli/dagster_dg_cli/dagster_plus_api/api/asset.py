"""Asset endpoints - REST-like interface."""

from typing import TYPE_CHECKING, Optional

from dagster_shared.plus.config import DagsterPlusCliConfig

from dagster_dg_cli.dagster_plus_api.graphql_adapter.asset import (
    get_dg_plus_api_asset_via_graphql,
    list_dg_plus_api_assets_via_graphql,
)

if TYPE_CHECKING:
    from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAsset, DgPlusApiAssetList


class DgPlusApiAssetAPI:
    def __init__(self, config: DagsterPlusCliConfig):
        self.config = config

    def list_assets(
        self,
        limit: Optional[int] = 50,
        cursor: Optional[str] = None,
    ) -> "DgPlusApiAssetList":
        """List assets with cursor-based pagination."""
        # Apply max limit constraint
        if limit and limit > 1000:
            limit = 1000

        return list_dg_plus_api_assets_via_graphql(self.config, limit=limit, cursor=cursor)

    def get_asset(self, asset_key: str) -> "DgPlusApiAsset":
        """Get single asset by slash-separated key (e.g., 'foo/bar')."""
        # Parse "foo/bar" to ["foo", "bar"]
        asset_key_parts = asset_key.split("/")

        return get_dg_plus_api_asset_via_graphql(self.config, asset_key_parts)
