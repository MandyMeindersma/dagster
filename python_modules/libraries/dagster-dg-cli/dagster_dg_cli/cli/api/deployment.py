"""Deployment API commands following GitHub CLI patterns."""

import sys

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper

from dagster_dg_cli.cli.api.formatters import format_deployments
from dagster_dg_cli.cli.api.shared import format_error_for_output, get_config_or_error
from dagster_dg_cli.dagster_plus_api.api.deployments import DgApiDeploymentApi


@click.command(name="list", cls=DgClickCommand, unlaunched=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@cli_telemetry_wrapper
def list_deployments_command(output_json: bool) -> None:
    """List all deployments in the organization."""
    config = get_config_or_error()
    api = DgApiDeploymentApi(config)

    try:
        deployments = api.list_deployments()
        output = format_deployments(deployments, as_json=output_json)
        click.echo(output)
    except Exception as e:
        error_output, exit_code = format_error_for_output(e, output_json)
        click.echo(error_output, err=True)
        sys.exit(exit_code)


@click.group(
    name="deployment",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "list": list_deployments_command,
    },
)
def deployment_group():
    """Manage deployments in Dagster Plus."""
