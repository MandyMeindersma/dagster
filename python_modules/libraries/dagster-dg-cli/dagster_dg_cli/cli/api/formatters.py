"""Output formatters for CLI display."""

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAsset, DgPlusApiAssetList
    from dagster_dg_cli.dagster_plus_api.schemas.deployment import DeploymentList


def format_deployments(deployments: "DeploymentList", as_json: bool) -> str:
    """Format deployment list for output."""
    if as_json:
        return deployments.model_dump_json(indent=2)

    lines = []
    for deployment in deployments.items:
        lines.extend(
            [
                f"Name: {deployment.name}",
                f"ID: {deployment.id}",
                f"Type: {deployment.type.value}",
                "",  # Empty line between deployments
            ]
        )

    return "\n".join(lines).rstrip()  # Remove trailing empty line


def format_assets(assets: "DgPlusApiAssetList", as_json: bool) -> str:
    """Format asset list for output."""
    if as_json:
        return assets.model_dump_json(indent=2)

    lines = []

    # Handle empty asset list
    if not assets.items:
        lines.append("No assets found.")
    else:
        for asset in assets.items:
            lines.extend(
                [
                    f"Asset Key: {asset.asset_key}",
                    f"ID: {asset.id}",
                    f"Group: {asset.group_name}",
                    f"Description: {asset.description or '(none)'}",
                    f"Kinds: {', '.join(asset.kinds) if asset.kinds else '(none)'}",
                    "",  # Empty line between assets
                ]
            )

    # Add pagination info
    if assets.cursor or assets.has_more:
        lines.extend(
            [
                "Pagination:",
                f"  Has More: {assets.has_more}",
            ]
        )
        if assets.cursor:
            lines.append(f"  Next Cursor: {assets.cursor}")
        lines.append("")

    return "\n".join(lines).rstrip()  # Remove trailing empty line


def format_asset(asset: "DgPlusApiAsset", as_json: bool) -> str:
    """Format single asset for output."""
    if as_json:
        return asset.model_dump_json(indent=2)

    lines = [
        f"Asset Key: {asset.asset_key}",
        f"ID: {asset.id}",
        f"Group: {asset.group_name}",
        f"Description: {asset.description or '(none)'}",
        f"Kinds: {', '.join(asset.kinds) if asset.kinds else '(none)'}",
    ]

    if asset.metadata_entries:
        lines.extend(["", "Metadata:"])
        for entry in asset.metadata_entries:
            label = entry.get("label", "")
            description = entry.get("description", "")
            lines.append(f"  {label}: {description}")

    return "\n".join(lines)
