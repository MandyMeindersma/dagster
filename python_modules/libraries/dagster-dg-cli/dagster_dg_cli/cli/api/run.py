"""Run API commands following GitHub CLI patterns."""

import json
from typing import Optional

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper
from dagster_shared.plus.config import DagsterPlusCliConfig
from dagster_shared.plus.config_utils import dg_api_options

# Lazy import to avoid loading pydantic at CLI startup
from dagster_dg_cli.cli.api.client import create_dg_api_graphql_client
from dagster_dg_cli.cli.api.formatters import format_run, format_runs


@click.command(name="list", cls=DgClickCommand, unlaunched=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@click.option(
    "--status",
    help="Filter by run status (e.g., SUCCESS, FAILURE, STARTED)",
)
@click.option(
    "--job",
    "job_name",
    help="Filter by job name",
)
@click.option(
    "--limit",
    type=int,
    help="Maximum number of runs to return",
)
@dg_api_options(deployment_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def list_runs_command(
    ctx: click.Context,
    output_json: bool,
    status: Optional[str],
    job_name: Optional[str],
    limit: Optional[int],
    organization: str,
    deployment: str,
    api_token: str,
    view_graphql: bool,
) -> None:
    """List runs in the deployment."""
    config = DagsterPlusCliConfig.create_for_deployment(
        organization=organization,
        deployment=deployment,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config, view_graphql=view_graphql)
    from dagster_dg_cli.api_layer.api.runs import DgApiRunApi

    api = DgApiRunApi(client)

    try:
        runs = api.list_runs(
            limit=limit,
            status=status,
            job_name=job_name,
        )
        output = format_runs(runs, as_json=output_json)
        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)
        raise click.ClickException(f"Failed to list runs: {e}")


@click.command(name="get", cls=DgClickCommand, unlaunched=True)
@click.argument("run_id", required=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@dg_api_options(deployment_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def get_run_command(
    ctx: click.Context,
    run_id: str,
    output_json: bool,
    organization: str,
    deployment: str,
    api_token: str,
    view_graphql: bool,
) -> None:
    """Get detailed information about a specific run."""
    config = DagsterPlusCliConfig.create_for_deployment(
        organization=organization,
        deployment=deployment,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config, view_graphql=view_graphql)
    from dagster_dg_cli.api_layer.api.runs import DgApiRunApi

    api = DgApiRunApi(client)

    try:
        run = api.get_run(run_id)
        output = format_run(run, as_json=output_json)
        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)
        raise click.ClickException(f"Failed to get run: {e}")


@click.group(
    name="run",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "list": list_runs_command,
        "get": get_run_command,
    },
)
def run_group():
    """Manage runs in Dagster Plus."""
