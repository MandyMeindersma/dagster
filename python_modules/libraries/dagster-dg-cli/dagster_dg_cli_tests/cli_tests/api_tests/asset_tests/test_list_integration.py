"""Integration tests for asset CLI commands.

These tests focus on CLI behavior and use minimal mocking of the GraphQL client
to test the integration between components.
"""

import json
from unittest.mock import patch

from click.testing import CliRunner
from dagster_dg_cli.cli.api.asset import list_assets_command, view_asset_command


class TestAssetListCommand:
    """Test the asset list CLI command integration."""

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.list_dg_plus_api_assets_via_graphql")
    def test_list_assets_text_output(self, mock_list_assets, mock_config, snapshot):
        """Test successful asset list with text output."""
        # Create mock asset data
        from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAsset, DgPlusApiAssetList

        mock_assets = DgPlusApiAssetList(
            items=[
                DgPlusApiAsset(
                    id="asset-1",
                    asset_key="users",
                    asset_key_parts=["users"],
                    description="User data table",
                    group_name="core_tables",
                    kinds=["dbt", "table"],
                    metadata_entries=[],
                ),
                DgPlusApiAsset(
                    id="asset-2",
                    asset_key="analytics/user_metrics",
                    asset_key_parts=["analytics", "user_metrics"],
                    description="Aggregated user metrics",
                    group_name="analytics",
                    kinds=["dbt", "view"],
                    metadata_entries=[],
                ),
            ],
            cursor="next_cursor",
            has_more=True,
        )

        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_list_assets.return_value = mock_assets

        # Run command
        runner = CliRunner()
        result = runner.invoke(list_assets_command, [])

        # Verify success
        assert result.exit_code == 0

        # Verify API was called with correct parameters
        mock_list_assets.assert_called_once()
        call_args = mock_list_assets.call_args[1]
        assert call_args["limit"] == 50  # default limit
        assert call_args["cursor"] is None

        # Snapshot the CLI output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.list_dg_plus_api_assets_via_graphql")
    def test_list_assets_json_output(self, mock_list_assets, mock_config, snapshot):
        """Test successful asset list with JSON output."""
        from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAsset, DgPlusApiAssetList

        mock_assets = DgPlusApiAssetList(
            items=[
                DgPlusApiAsset(
                    id="asset-1",
                    asset_key="users",
                    asset_key_parts=["users"],
                    description="User data table",
                    group_name="core_tables",
                    kinds=["dbt", "table"],
                    metadata_entries=[{"label": "owner", "description": "data-team"}],
                )
            ],
            cursor=None,
            has_more=False,
        )

        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_list_assets.return_value = mock_assets

        # Run command with --json flag
        runner = CliRunner()
        result = runner.invoke(list_assets_command, ["--json"])

        # Verify success
        assert result.exit_code == 0

        # Parse and snapshot the JSON output to avoid formatting differences
        actual_output = json.loads(result.output)
        snapshot.assert_match(actual_output)

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.list_dg_plus_api_assets_via_graphql")
    def test_list_assets_with_pagination_options(self, mock_list_assets, mock_config, snapshot):
        """Test asset list with pagination parameters."""
        from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAssetList

        mock_assets = DgPlusApiAssetList(items=[], cursor=None, has_more=False)

        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_list_assets.return_value = mock_assets

        # Run command with pagination options
        runner = CliRunner()
        result = runner.invoke(list_assets_command, ["--limit", "10", "--cursor", "abc123"])

        # Verify success
        assert result.exit_code == 0

        # Verify API was called with correct parameters
        mock_list_assets.assert_called_once()
        call_args = mock_list_assets.call_args[1]
        assert call_args["limit"] == 10
        assert call_args["cursor"] == "abc123"

        # Snapshot the CLI output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.list_dg_plus_api_assets_via_graphql")
    def test_list_assets_api_error(self, mock_list_assets, mock_config, snapshot):
        """Test handling of API errors."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_list_assets.side_effect = Exception("GraphQL API Error")

        # Run command
        runner = CliRunner()
        result = runner.invoke(list_assets_command, [])

        # Verify error handling
        assert result.exit_code != 0

        # Snapshot the error output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.list_dg_plus_api_assets_via_graphql")
    def test_list_assets_api_error_json_output(self, mock_list_assets, mock_config, snapshot):
        """Test handling of API errors with JSON output."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_list_assets.side_effect = Exception("GraphQL API Error")

        # Run command with --json flag
        runner = CliRunner()
        result = runner.invoke(list_assets_command, ["--json"])

        # Verify error handling
        assert result.exit_code != 0

        # Snapshot the error output (should be JSON format)
        snapshot.assert_match(result.output)


class TestAssetViewCommand:
    """Test the asset view CLI command integration."""

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.get_dg_plus_api_asset_via_graphql")
    def test_view_asset_text_output(self, mock_get_asset, mock_config, snapshot):
        """Test successful asset view with text output."""
        from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAsset

        mock_asset = DgPlusApiAsset(
            id="asset-1",
            asset_key="users",
            asset_key_parts=["users"],
            description="User data table",
            group_name="core_tables",
            kinds=["dbt", "table"],
            metadata_entries=[
                {"label": "owner", "description": "data-team"},
                {"label": "rows", "description": "1000000", "intValue": 1000000},
            ],
        )

        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_get_asset.return_value = mock_asset

        # Run command
        runner = CliRunner()
        result = runner.invoke(view_asset_command, ["users"])

        # Verify success
        assert result.exit_code == 0

        # Verify API was called with correct asset key
        mock_get_asset.assert_called_once()
        call_args = mock_get_asset.call_args[0]
        assert call_args[1] == ["users"]  # asset_key_parts

        # Snapshot the CLI output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.get_dg_plus_api_asset_via_graphql")
    def test_view_asset_json_output(self, mock_get_asset, mock_config, snapshot):
        """Test successful asset view with JSON output."""
        from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAsset

        mock_asset = DgPlusApiAsset(
            id="asset-2",
            asset_key="analytics/user_metrics",
            asset_key_parts=["analytics", "user_metrics"],
            description="Aggregated user metrics",
            group_name="analytics",
            kinds=["dbt", "view"],
            metadata_entries=[],
        )

        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_get_asset.return_value = mock_asset

        # Run command with --json flag
        runner = CliRunner()
        result = runner.invoke(view_asset_command, ["analytics/user_metrics", "--json"])

        # Verify success
        assert result.exit_code == 0

        # Verify API was called with correct nested asset key
        mock_get_asset.assert_called_once()
        call_args = mock_get_asset.call_args[0]
        assert call_args[1] == ["analytics", "user_metrics"]

        # Parse and snapshot the JSON output
        actual_output = json.loads(result.output)
        snapshot.assert_match(actual_output)

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.get_dg_plus_api_asset_via_graphql")
    def test_view_asset_not_found(self, mock_get_asset, mock_config, snapshot):
        """Test handling when asset is not found."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_get_asset.side_effect = Exception("Asset not found: nonexistent-asset")

        # Run command
        runner = CliRunner()
        result = runner.invoke(view_asset_command, ["nonexistent-asset"])

        # Verify error handling
        assert result.exit_code != 0

        # Snapshot the error output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.asset.DagsterPlusCliConfig.create_for_deployment")
    @patch("dagster_dg_cli.dagster_plus_api.api.asset.get_dg_plus_api_asset_via_graphql")
    def test_view_asset_not_found_json_output(self, mock_get_asset, mock_config, snapshot):
        """Test handling when asset is not found with JSON output."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_get_asset.side_effect = Exception("Asset not found: nonexistent-asset")

        # Run command with --json flag
        runner = CliRunner()
        result = runner.invoke(view_asset_command, ["nonexistent-asset", "--json"])

        # Verify error handling
        assert result.exit_code != 0

        # Snapshot the error output (should be JSON format)
        snapshot.assert_match(result.output)
