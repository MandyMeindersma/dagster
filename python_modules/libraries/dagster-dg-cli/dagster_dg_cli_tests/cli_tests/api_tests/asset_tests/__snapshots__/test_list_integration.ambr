# serializer version: 1
# name: TestAssetListCommand.test_list_assets_api_error[0]
  '''
  Error querying Dagster Plus API: GraphQL API Error
  Error: Failed to list assets: GraphQL API Error
  
  '''
# ---
# name: TestAssetListCommand.test_list_assets_api_error_json_output[0]
  '''
  {"error": "GraphQL API Error"}
  Error: Failed to list assets: GraphQL API Error
  
  '''
# ---
# name: TestAssetListCommand.test_list_assets_json_output[0]
  dict({
    'cursor': None,
    'has_more': False,
    'items': list([
      dict({
        'asset_key': 'users',
        'asset_key_parts': list([
          'users',
        ]),
        'description': 'User data table',
        'group_name': 'core_tables',
        'id': 'asset-1',
        'kinds': list([
          'dbt',
          'table',
        ]),
        'metadata_entries': list([
          dict({
            'description': 'data-team',
            'label': 'owner',
          }),
        ]),
      }),
    ]),
  })
# ---
# name: TestAssetListCommand.test_list_assets_text_output[0]
  '''
  Asset Key: users
  ID: asset-1
  Group: core_tables
  Description: User data table
  Kinds: dbt, table
  
  Asset Key: analytics/user_metrics
  ID: asset-2
  Group: analytics
  Description: Aggregated user metrics
  Kinds: dbt, view
  
  Pagination:
    Has More: True
    Next Cursor: next_cursor
  
  '''
# ---
# name: TestAssetListCommand.test_list_assets_with_pagination_options[0]
  '''
  No assets found.
  
  '''
# ---
# name: TestAssetViewCommand.test_view_asset_json_output[0]
  dict({
    'asset_key': 'analytics/user_metrics',
    'asset_key_parts': list([
      'analytics',
      'user_metrics',
    ]),
    'description': 'Aggregated user metrics',
    'group_name': 'analytics',
    'id': 'asset-2',
    'kinds': list([
      'dbt',
      'view',
    ]),
    'metadata_entries': list([
    ]),
  })
# ---
# name: TestAssetViewCommand.test_view_asset_not_found[0]
  '''
  Error querying Dagster Plus API: Asset not found: nonexistent-asset
  Error: Failed to get asset: Asset not found: nonexistent-asset
  
  '''
# ---
# name: TestAssetViewCommand.test_view_asset_not_found_json_output[0]
  '''
  {"error": "Asset not found: nonexistent-asset"}
  Error: Failed to get asset: Asset not found: nonexistent-asset
  
  '''
# ---
# name: TestAssetViewCommand.test_view_asset_text_output[0]
  '''
  Asset Key: users
  ID: asset-1
  Group: core_tables
  Description: User data table
  Kinds: dbt, table
  
  Metadata:
    owner: data-team
    rows: 1000000
  
  '''
# ---
